helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ZkMetadataStoreDirectory.java

```diff
@@ -155,6 +159,19 @@ private void init(String namespace, String zkAddress) throws InvalidRoutingDataE
     return routingData;
   }
 
+  @Override
+  public boolean setNamespaceRoutingData(String namespace, Map<String, List<String>> routingData) {
+    synchronized (this) {
+      if (!_routingDataWriterMap.containsKey(namespace)) {
```

Comment:
This is like double lock checking. You may also want to check it before synchronized which also improves performance.
Considering namespace is not changed, you don’t have to check it in within synchronized block.

Reply:
Sure.